File: src/repositories/blog_repository.rs
use diesel::prelude::*;
use diesel::sqlite::SqliteConnection;

use crate::models::blog::{Blog, UpdateBlogDto};
use crate::repositories::base_repository::{
    CreateDTO, Entity, Repository, RepositoryError, UpdateDTO,
};
use crate::schema::blogs;

pub struct BlogRepository {
    connection: SqliteConnection,
}

impl BlogRepository {
    pub fn new(connection: SqliteConnection) -> Self {
        Self { connection }
    }
}

impl Repository<Blog> for BlogRepository {
    fn find_all(&mut self) -> Result<Vec<Blog>, RepositoryError> {
        todo!()
    }
    fn create(&mut self, entity: &Blog) -> Result<(), RepositoryError> {
        todo!()
    }
    fn delete(&mut self, id: &<Blog as Entity>::Id) -> Result<usize, RepositoryError> {
        todo!()
    }
    fn find_by_id(&mut self, id: &<Blog as Entity>::Id) -> Result<Blog, RepositoryError> {
        todo!()
    }
    fn update(
        &mut self,
        id: &<Blog as Entity>::Id,
        dto: &dyn UpdateDTO,
    ) -> Result<Blog, RepositoryError> {
        todo!()
    }
}

File: src/repositories/BlogRepository.ts
import { db } from '../db';
import { blogs } from '../db/schema';
import { eq } from 'drizzle-orm';
import type { Blog, NewBlog } from '../models/Blog';
import type { IRepository } from '../interfaces/IRepository';

export class BlogRepository implements IRepository<Blog, NewBlog> {
  async getAll(): Promise<Blog[]> {
    return db.select().from(blogs);
  }

  async getById(id: string): Promise<Blog | undefined> {
    const result = await db.select().from(blogs).where(eq(blogs.id, id));
    return result[0];
  }

  async create(data: NewBlog): Promise<Blog> {
    const result = await db.insert(blogs).values(data).returning();
    return result[0];
  }

  async delete(id: string): Promise<void> {
    await db.delete(blogs).where(eq(blogs.id, id));
  }

  async update(id: string, data: Partial<NewBlog>): Promise<Blog | undefined> {
    const result = await db
      .update(blogs)
      .set({ ...data })
      .where(eq(blogs.id, id))
      .returning();
    return result[0];
  }
}
